showkase-processor/src/main/java/com/airbnb/android/showkase/processor/writer/PaparazziShowkaseScreenshotTestWriter.kt

```diff
@@ -0,0 +1,268 @@
+package com.airbnb.android.showkase.processor.writer
+
+import androidx.room.compiler.processing.XFiler
+import androidx.room.compiler.processing.XProcessingEnv
+import androidx.room.compiler.processing.writeTo
+import com.squareup.kotlinpoet.AnnotationSpec
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.FunSpec
+import com.squareup.kotlinpoet.KModifier
+import com.squareup.kotlinpoet.LIST
+import com.squareup.kotlinpoet.ParameterSpec
+import com.squareup.kotlinpoet.ParameterizedTypeName
+import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.TypeSpec
+
+class PaparazziShowkaseScreenshotTestWriter(private val environment: XProcessingEnv) {
+    @Suppress("LongParameterList")
+    internal fun generateScreenshotTests(
+        screenshotTestPackageName: String,
+        rootModulePackageName: String,
+        testClassName: String
+    ) {
+        val showkaseScreenshotTestClassName = "${testClassName}_PaparazziShowkaseCodegen"
+        val fileBuilder = getFileBuilder(screenshotTestPackageName, showkaseScreenshotTestClassName)
+        fileBuilder
+            .addImport(rootModulePackageName, "getMetadata")
+            .addType(
+                with(TypeSpec.classBuilder(showkaseScreenshotTestClassName)) {
+                    superclass(ClassName(screenshotTestPackageName, testClassName))
+                    addAnnotation(
+                        AnnotationSpec.builder(ShowkaseScreenshotTestWriter.RUNWITH_CLASSNAME)
+                            .addMember("%T::class", TEST_PARAMETER_INJECTOR_CLASSNAME)
+                            .build()
+                    )
+                    addProperty(addPaparazziTestRuleProperty())
+                    addPreviewProvider()
+                    addProvider(
+                        "PaparazziShowkaseDeviceConfigProvider",
+                        LIST.parameterizedBy(DEVICE_CONFIG_CLASS_NAME),
+                        "deviceConfigs"
+                    )
+                    addProvider(
+                        "PaparazziShowkaseLayoutDirectionProvider",
+                        LIST.parameterizedBy(LAYOUT_DIRECTION_CLASS_NAME),
+                        "layoutDirections"
+                    )
+                    addProvider(
+                        "PaparazziShowkaseUIModeProvider",
+                        LIST.parameterizedBy(UI_MODE_CLASS_NAME),
+                        "uiModes"
+                    )
+                    addTest()
+                    build()
+                }
+            )
+
+        fileBuilder.build().writeTo(environment.filer, mode = XFiler.Mode.Aggregating)
+    }
+
+    private fun TypeSpec.Builder.addPreviewProvider() {
```

Comment:
Is it necessary to code gen this (and some of the other objects?) They seem purely static from what I can see

Reply:
@BenSchwab That was necessary so that users could override the behavior of some of the methods declared inside the `CompanionObject` interface. If I were to move these objects directly inside the interface, I'd need to change the interface to an abstract class, but at that point, I won't be able to allow overriding the behavior which is vital. Is there another way to structure this? 🤔